### DISCUSSION TITLE
Push or pull on ghcr package results in error when authenticating with an oauth app

### DISCUSSION THREAD

**Discussion by logikaljay** [community/community#26921] [Packages]
<p>This is a strange one that maybe someone has experienced before.</p>
<p>When I generated a PAT with the following scopes: <code>repo read:packages write:packages delete:packages</code> and use that PAT in <code>docker login ghcr.io -u logikaljay</code> I am able to push and pull packages without any problems.</p>
<p>When I use the OAuth app with either device flow or auth code flow, with the same scopes, I get the following errors.<br>
<em>On push</em>:</p>
<pre><code class="lang-auto">$ docker push ghcr.io/logikaljay/test-fn:latest
The push refers to repository [ghcr.io/logikaljay/test-fn]
82c238761b57: Retrying in 1 second 
1049296cf531: Retrying in 1 second 
c3c26b644eb6: Retrying in 1 second 
c69b14870e0e: Retrying in 1 second 
999e8d6972a7: Retrying in 1 second 
...
77cae8ab23bf: Waiting 
unauthorized
</code></pre>
<p><em>On pull</em>:</p>
<pre><code class="lang-auto">$ docker pull ghcr.io/logikaljay/test-fn:latest
Error response from daemon: pull access denied for ghcr.io/logikaljay/test-fn, repository does not exist or may require 'docker login': denied: permission_denied: The token provided does not match expected scopes.
</code></pre>
<p>After authenticating with device flow, the response from <code>https://github.com/login/oauth/access_token</code> is (access_token obfuscated):</p>
<pre><code class="lang-auto">{
  access_token: '***',
  token_type: 'bearer',
  scope: 'delete:packages,read:packages,repo,write:packages'
}
</code></pre>
<p>Has anyone had any luck trying to push a docker image to ghcr using an OAuth app based access_token instead of a manually generated PAT ?</p>

**Comment by logikaljay** [comment, marked as answer]
<p>For other people that may find this topic due to keyword searches, I believe it is answered in another topic:</p>
<aside class="quote group-github-employees" data-username="clarkbw" data-post="2" data-topic="130666" data-full="true">
<div class="title">
<div class="quote-controls"></div>
<a href="https://github.community/t/how-to-use-installation-access-token-in-ghcr-io-authorization/130666/2">How to use Installation Access Token in ghcr.io authorization?</a>
</div>
<blockquote>
<p>GHCR can’t accept App tokens, only PATs for now. We’re working on a solution to allow for the Actions <code>GITHUB_TOKEN</code> and then could look into this after.</p>
</blockquote>
</aside>

**Reply by ipatch** [reply]
that link is redirecting to the discussions index.

**Reply by oknoorap** [reply]
this link doesn't work anymore

**Reply by ipatch** [reply]
> this link doesn't work anymore

i think this is about as good as you'll get,

https://web.archive.org/web/*/https://github.community/t/how-to-use-installation-access-token-in-ghcr-io-authorization/130666*

**Reply by fxnn** [reply]
TL;DR: After the request from 2020-09-08, they said they're "working on a solution to allow for the Actions `GITHUB_TOKEN` and then could look into this after".

On 2021-03-30, they [actually released the `GITHUB_TOKEN` support](https://github.blog/changelog/2021-03-24-packages-container-registry-now-supports-github_token/). 

Whether GitHub Apps were eventually worked on remains unknown, because the rest of the conversation is not archived, unfortunately. The last message regarding this was that "App tokens are in the backlog but I don’t have a timeline yet" (2021-02-18).

**Reply by CermDevops** [reply]
As of today, GitHub Apps are still unable to push to the GHCR.
Got to use the ${{ github.token }} instead.
